man/ssi-cgi.1
.\" Hiawatha manualpage
.\"
.TH SSI-CGI 1

.SH NAME
ssi-cgi - Server Side Includes parser as a CGI program


.SH SYNOPSIS
.B ssi-cgi
.br
-h: show help and exit.
.br
-v: show version and compile information and exit.


.SH DESCRIPTION
SSI parser which works like a CGI application.


.SH SSI COMMANDS
SSI commands have the following format: <!--#command parameter="value"-->. This chapter shows all the available SSI commands and their parameters in ssi-cgi.
.TP
.B config
Use the config command to control the output.
.br

Parameters:
.br
.B	errmsg:
Set the message ssi-cgi must print in case of an error.
.br

Example: <!--#config errmsg="SSI error"-->
.TP
.B echo
Print the value of a variable.
.br

Parameters:
.br
.B	var:
Print the value of an environment variable.
.br

Example: <!--#echo var="DOCUMENT_ROOT"-->
.TP
.B exec
Execute a program and include its output at the position of the exec command.
.br

Parameters:
.br
.B	cgi:
Execute a CGI program relative to the website root.
.br
.B	cmd:
Execute any command as a parameter of /bin/sh.
.br

Example: <!--#exec cmd="ls -Flsh"-->
.TP
.B fsize
Print the size of a file.
.br

Parameters:
.br
.B	file:
The file of which the size should be printed.
.br

Example: <!--#fsize file="info.txt"-->
.TP
.B include
Include another file at the position of the include command.
.br

Parameters:
.br
.B	file:
Include another file relative to the current file.
.br
.B	virtual:
Include another file relative to the website root.
.br

Example: <!--#include file="information.txt"-->
.TP
.B printenv
Print all environment variables.
.br

Example: <!--#printenv-->


.SH SEE ALSO
SSI-CGI is part of the Hiawatha webserver. See hiawatha(1) for more information about Hiawatha.


.SH AUTHOR
Hugo Leisink <hugo@hiawatha-webserver.org> - \fIhttp://www.hiawatha-webserver.org/\fP


ChangeLog
hiawatha (8.6) stable; urgency=low

  * PolarSSL updated to version 1.2. Added support for TLS 1.2 and
    secure renegotiation.
  * Added support for Server Name Indication.
  * MinSSLversion option added.
  * ServerRoot option removed.
  * Improved MacOS X package building script.
  * Marked php-fcgi as deprecated. Use php-fpm instead.
  * Small bugfixes and improvements.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 31 Oct 2012 19:10:32 +0100

hiawatha (8.5) stable; urgency=low

  * Improved Reverse Proxy.
  * Changed error message style.
  * Renamed Command Channel to Tomahawk.
  * Return 403 instead of 401 upon correct password for HTTP
    authentication but user not in right group.
  * Small improvements.
  * Bugfix: replaced select() with poll() to prevent crashes in case of
    large amount of simultaneous connections. Thanks to Peter Bex.

 -- Hugo Leisink <hugo@leisink.net>  Sun,  9 Sep 2012 11:39:12 +0200

hiawatha (8.4) stable; urgency=low

  * MaxServerLoad option added.
  * PolarSSL updated to version 1.1.4.
  * Small bugfixes and improvements.
  * Bugfix: invalid reverse proxy request when URL parameters are present.

 -- Hugo Leisink <hugo@leisink.net>  Thu,  7 Jun 2012 20:07:46 +0200

hiawatha (8.3.2) stable; urgency=high

  * Bugfix: memory leak in SSL library.

 -- Hugo Leisink <hugo@leisink.net>  Tue, 29 May 2012 18:02:59 +0200

hiawatha (8.3.1) stable; urgency=low

  * Improved security for reverse proxy (works with PreventSQLi, etc).

 -- Hugo Leisink <hugo@leisink.net>  Mon, 28 May 2012 21:50:31 +0200

hiawatha (8.3) stable; urgency=low

  * ReverseProxy option added.
  * PolarSSL updated to version 1.1.3.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 23 May 2012 18:11:56 +0200

hiawatha (8.2) stable; urgency=low

  * WebDAVapp option added. Enables support for WebDAV applications
    like ownCloud (http://owncloud.org/).
  * Removed support for the OPTIONS method.
  * AllowDotFiles option added.
  * Global forks setting in php-fcgi.conf moved to Server setting.
  * Small bugfixes and improvements.

 -- Hugo Leisink <hugo@leisink.net>  Tue,  1 May 2012 17:48:27 +0200

hiawatha (8.1) stable; urgency=low

  * BanOnInvalidURL option added.
  * PolarSSL updated to version 1.1.1.
  * Small improvements in Windows packaging script.
  * Bugfix: paths missing in default values and examples in manual pages.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 25 Feb 2012 19:02:41 +0100

hiawatha (8.0) stable; urgency=low

  * Replaced Autoconf with CMake. Many thanks to Sander Niemeijer.
  * Replaced OpenSSL with PolarSSL. Many thanks to Paul Bakker.
  * AllowedCiphers and DHparameters options removed.
  * Added IE7 to UrlToolkit's OldBrowser list, removed IE5.
  * MaxUrlLength option added, can return 414 Request-URI Too Long.
  * Changed default value of TriggerOnCGIstatus to 'no'.
  * Equalized format of logfiles.
  * Extra checks added to php-fcgi.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Fri, 27 Jan 2012 12:06:10 +0100

hiawatha (7.8.2) stable; urgency=high

  * Improved SQL injection detection.
  * Bugfix: memory leak in PreventSQLi routine.
  * Bugfix: potential server freeze with 100% CPU in CGI output caching.

 -- Hugo Leisink <hugo@leisink.net>  Fri, 18 Nov 2011 06:51:07 +0100

hiawatha (7.8.1) stable; urgency=low

  * Small bugfixes and improvements.
  * Bugfix: null byte in HTTP header of cached CGI content.

 -- Hugo Leisink <hugo@leisink.net>  Wed,  9 Nov 2011 17:21:52 +0100

hiawatha (7.8) stable; urgency=low

  * Control CGI output cache via X-Hiawatha-Cache and X-Hiawatha-Cache-Remove
    CGI headers. See the CGI OUTPUT CACHE section in the manual page.
  * BanOnWrongPassword now also triggers on wrong username.
  * Small improvements.
  * Bugfix: timeout issue with large POST requests on SSL connections.

 -- Hugo Leisink <hugo@leisink.net>  Mon, 31 Oct 2011 21:27:18 +0100

hiawatha (7.7) stable; urgency=low

  * First parameter of Alias can now contain subdirectories.
  * Improved stability for connections with SSL client authentication.
  * Bugfix: BanOnFlooding was broken.

 -- Hugo Leisink <hugo@leisink.net>  Tue,  4 Oct 2011 19:48:30 +0200

hiawatha (7.6) stable; urgency=low

  * PreventSQLi option rewritten.

 -- Hugo Leisink <hugo@leisink.net>  Sun, 21 Aug 2011 08:06:21 +0200

hiawatha (7.5) stable; urgency=low

  * OldBrowser option added to URL toolkit.
  * Improved mimetype configuration.
  * Do-not-track HTTP header support.
  * Password file entries can now be created with Wigwam.
  * Small bugfixes and improvements.
  * Bugfix: sent one byte too few for Range -XX.
  * Bugfix: possible crash when using PreventSQLi.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 28 May 2011 15:39:13 +0200

hiawatha (7.4.1) stable; urgency=high

  * Bugfix: integer overflow in fetch_request() which could
    lead to a server crash.

 -- Hugo Leisink <hugo@leisink.net>  Sat, 26 Feb 2011 10:32:24 +0100

hiawatha (7.4) stable; urgency=medium

  * Connections per IP added to RequestLimitMask.
  * NoExtensionAs made a per-host setting.
  * Small bugfixes and improvements.
  * Bugfix: usage of HideProxy caused Hiawatha to refuse new connections
    after ConnectionsTotal connections.
  * Bugfix: memory leak in XSLT module.

 -- Hugo Leisink <hugo@leisink.net>  Mon,  8 Nov 2010 20:58:54 +0100

hiawatha (7.3) stable; urgency=low

  * RequestLimitMask option added.
  * URL parameters for ErrorHandler.
  * Support for Haiku OS.
  * Small security bugfixes.

 -- Hugo Leisink <hugo@leisink.net>  Sun,  6 Jun 2010 23:18:37 +0200

hiawatha (7.2) stable; urgency=low

  * URL toolkit code restructured.
  * UseSSL option added to URL toolkit.
  * Digest HTTP authentication works with htdigest(1) created password files.
  * Small improvements.

 -- Hugo Leisink <hugo@leisink.net>  Wed, 21 Apr 2010 18:12:37 +0200

hiawatha (7.1) stable; urgency=low

  * Small bugfixes.
  * Bugfix: deny access and redirect result via toolkit subroutine.
  * Bugfix: broken flooding protection.


src/ssi-cgi.c
/* This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License. For a copy,
 * see http://www.gnu.org/licenses/gpl-2.0.html.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 */

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <poll.h>
#include <time.h>
#include <errno.h>
#include <fcntl.h>
#include "global.h"
#include "alternative.h"
#include "liblist.h"
#include "libstr.h"
#include "libfs.h"

#define CGI_BUFFER_SIZE    4096
#define MAX_LINE_SIZE      1024
#define MAX_CWD             256
#define MAX_INCLUDE_DEPTH    16
#define SSI_EXTENSIONS     "shtml, stm, shtm"

typedef struct type_parameter {
	char *key, *value;
	struct type_parameter *next;
} t_parameter;

typedef struct {
	t_charlist extension;
	int  include_depth;
	char *document_root;
	char *working_dir;
	char *errmsg;
} t_config;

typedef enum { cgi, cmd } t_exec_type;

static char *CURRENT_DIRECTORY = ".";

int process_ssi_file(t_config *config, char *file);

/* Get configuration
 */
t_config *get_configuration(void) {
	t_config *config;
	char *extensions;

	if ((config = (t_config*)malloc(sizeof(t_config))) == NULL) {
		return NULL;
	}

	if ((extensions = strdup(SSI_EXTENSIONS)) == NULL) {
		return NULL;
	} else if (parse_charlist(extensions, &(config->extension)) == -1) {
		return NULL;
	}

	config->include_depth = 0;

	if ((config->document_root = getenv("DOCUMENT_ROOT")) == NULL) {
		config->document_root = CURRENT_DIRECTORY;
	}

	if ((config->working_dir = (char*)malloc(MAX_CWD)) == NULL) {
		return NULL;
	} else if (getcwd(config->working_dir, MAX_CWD) == NULL) {
		config->working_dir = CURRENT_DIRECTORY;
	}

	config->errmsg = NULL;

	return config;
}

/* Close pipe
 */
void close_pipe_end(int *pipe_end, int *count) {
	if (*pipe_end != -1) {
		close(*pipe_end);
		(*pipe_end) = -1;
		(*count)--;
	}
}

/* Get parameter value based in key
 */
char *get_parameter(t_parameter *param, char *key) {
	while (param != NULL) {
		if (strcmp(param->key, key) == 0) {
			return param->value;
		}
		param = param->next;
	}

	return NULL;
}

/* Remove parameters
 */
void *remove_parameters(t_parameter *param) {
	t_parameter *delete;

	while (param != NULL) {
		delete = param;
		param = param->next;

		free(delete);
	}

	return NULL;
}

/* Print error message
 */
void print_error(t_config *config, char *message) {
	if (config != NULL) {
		if (config->errmsg != NULL) {
			printf("%s\n", config->errmsg);
		}
	}

	fprintf(stderr, "%s\n", message);
}

/* Execute file
 */
int execute_command(char *command, t_exec_type type) {
	int output_pipe[2], error_pipe[2], result, skip, poll_size;
	int bytes_read, bytes_in_buffer = 0, open_pipes = 2;
	char buffer[CGI_BUFFER_SIZE], *pos;
	bool header_read;
	struct pollfd poll_data[2];
	pid_t cgi_pid;

	fflush(stdout);
	fflush(stderr);

	header_read = (type == cmd);

	if (pipe(output_pipe) == -1) {
		return -1;
	} else if (pipe(error_pipe) == -1) {
		close(output_pipe[0]);
		close(output_pipe[1]);
		return -1;
	}

	if ((cgi_pid = fork()) == -1) {
		return -1;
	}

	if (cgi_pid == 0) {
		/* Child. Executes command.
		 */
		dup2(output_pipe[1], STDOUT_FILENO);
		dup2(error_pipe[1], STDERR_FILENO);

		close(output_pipe[0]);
		close(output_pipe[1]);
		close(error_pipe[0]);
		close(error_pipe[1]);

		fcntl(STDOUT_FILENO, F_SETFD, 0);
		fcntl(STDERR_FILENO, F_SETFD, 0);

		if (type == cgi) {
			if ((pos = strchr(command, '?')) != NULL) {
				*(pos++) = '\0';
				setenv("QUERY_STRING", pos, 1);
			}
			url_decode(command);

			execlp(command, command, (char*)NULL);
		} else {
			if ((result = system(command)) != -1) {
				exit(result);
			}
		}

		exit(EXIT_FAILURE);
	}

	/* Parent. Reads command output.
	 */
	close(output_pipe[1]);
	close(error_pipe[1]);

	poll_size = 0;
	if (output_pipe[0] != -1) {
		poll_data[poll_size].fd = output_pipe[0];
		poll_data[poll_size].events = POLL_EVENT_BITS;
		poll_size++;
	}
	if (error_pipe[0] != -1) {
		poll_data[poll_size].fd = error_pipe[0];
		poll_data[poll_size].events = POLL_EVENT_BITS;
		poll_size++;
	}

	if (poll_size == 0) {
		return -1;
	}

	do {
		switch (poll((struct pollfd*)&poll_data, poll_size, 1000)) {
			case -1:
				return -1;
			case 0:
				break;
			default:
				/* Output
				 */
				if ((output_pipe[0] != -1) && (poll_data[0].revents != 0)) {
					bytes_read = read(output_pipe[0], buffer + bytes_in_buffer, CGI_BUFFER_SIZE - bytes_in_buffer);
					switch (bytes_read) {
						case -1:
							return -1;
						case 0:
							close_pipe_end(&output_pipe[0], &open_pipes);
							header_read = true;
							break;
						default:
							if (header_read == false) {
								bytes_in_buffer += bytes_read;

								if ((pos = strstr(buffer, "\r\n\r\n")) == NULL) {
									pos = strstr(buffer, "\n\n");
									skip = 2;
								} else {
									skip = 4;
								}

								if (pos != NULL) {
									/* Skip CGI header
									 */
									pos += skip;
									if (write_buffer(STDOUT_FILENO, pos, bytes_in_buffer - (pos - buffer)) == -1) {
										return -1;
									}
									header_read = true;
									bytes_in_buffer = 0;
								}
							} else if (write_buffer(STDOUT_FILENO, buffer, bytes_read) == -1) {
								return -1;
							}
					}
				}

				/* Error
				 */
				if ((error_pipe[0] != -1) && header_read && (poll_data[1].revents != 0)) {
					bytes_read = read(error_pipe[0], buffer, CGI_BUFFER_SIZE);
					switch (bytes_read) {
						case -1:
							return -1;
						case 0:
							close_pipe_end(&error_pipe[0], &open_pipes);
							break;
						default:
							if (write_buffer(STDERR_FILENO, buffer, bytes_read) == -1) {
								return -1;
							}
					}
				}
		}
	} while (open_pipes > 0);

	waitpid(cgi_pid, NULL, 0);

	return 0;
}

/* Include file
 */
int include_file(t_config *config, char *directory, char *file) {
	FILE *fp;
	char line[MAX_LINE_SIZE + 1], *extension;

	if (chdir(directory) == -1) {
		return -1;
	}

	if ((extension = strrchr(file, '.')) != NULL) {
		extension++;
		if (in_charlist(extension, &(config->extension))) {
			if (config->include_depth >= MAX_INCLUDE_DEPTH) {
				print_error(config, "maximum include depth reached");
			} else {
				config->include_depth++;
				if (process_ssi_file(config, file) == -1) {
					print_error(config, "|");
					return -1;
				}
				config->include_depth--;
			}

			return 0;
		}
	}

	if ((fp = fopen(file, "r")) == NULL) {
		return -1;
	}

	line[MAX_LINE_SIZE] = '\0';
	while (fgets(line, MAX_LINE_SIZE, fp) != NULL) {
		printf("%s", line);
	}

	fclose(fp);

	return 0;
}

/* Execute SSI command
 */
int execute_ssi_command(t_config *config, char *command, t_parameter *param) {
	extern char **environ;
	char *str, **env;
	struct stat status;

	if (strcmp(command, "config") == 0) {
		/* CONFIG
		 */
		while (param != NULL) {
			/* errmsg
			 */
			if (strcmp(param->key, "errmsg") == 0) {
				if (config->errmsg != NULL) {
					free(config->errmsg);
				}
				if ((config->errmsg = strdup(param->value)) == NULL) {
					print_error(config, "strdup error");
				}
			}

			param = param ->next;
		}
	} else if (strcmp(command, "echo") == 0) {
		/* ECHO
		 */
		while (param != NULL) {
			/* var
			 */
			if (strcmp(param->key, "var") == 0) {
				if ((str = getenv(param->value)) != NULL) {
					printf("%s", str);
				}
			}

			param = param->next;
		}
	} else if (strcmp(command, "exec") == 0) {
		/* EXEC
		 */
		while (param != NULL) {
			if (strcmp(param->key, "cgi") == 0) {
				execute_command(param->value, cgi);
			} else if (strcmp(param->key, "cmd") == 0) {
				execute_command(param->value, cmd);
			}

			param = param->next;
		}
	} else if (strcmp(command, "fsize") == 0) {
		/* FSIZE
		 */
		while (param != NULL) {
			if (strcmp(param->key, "file") == 0) {
				if (stat(param->value, &status) == 0) {
					printf("%ld", (long)status.st_size);
				} else {
					print_error(config, "stat error");
				}
			}

			param = param->next;
		}
	} else if (strcmp(command, "include") == 0) {
		/* INCLUDE
		 */
		while (param != NULL) {
			if (strcmp(param->key, "file") == 0) {
				/* file
				 */
				if (chdir(config->working_dir) == -1) {
					print_error(config, "chdir error");
				} else if (include_file(config, CURRENT_DIRECTORY, param->value) == -1) {
					print_error(config, "include error");
				}
			} else if (strcmp(param->key, "virtual") == 0) {
				/* virtual
				 */
				while (*param->value == '/') {
					param->value++;
				}

				if (chdir(config->document_root) == -1) {
					print_error(config, "chdir error");
				} else if (include_file(config, CURRENT_DIRECTORY, param->value) == -1) {
					print_error(config, "include error");
				}
			}

			param = param->next;
		}
	} else if (strcmp(command, "printenv") == 0) {
		/* PRINTENV
		 */
		env = environ;
		while (*env != NULL) {
			printf("%s\n", *env);
			env++;
		}
	}

	return 0;
}

/* Process SSI file
 */
int process_ssi_file(t_config *config, char *file) {
	FILE *fp;
	char line[MAX_LINE_SIZE + 1], *scan, *tag_begin, *tag_end;
	char *command, *param_key, *param_value, *param_end;
	t_parameter *param = NULL, *last = NULL, *new;
	bool error_found;

	if ((fp = fopen(file, "r")) == NULL) {
		fprintf(stderr, "file '%s' not found", file);
		return -1;
	}

	line[MAX_LINE_SIZE] = '\0';
	while (fgets(line, MAX_LINE_SIZE, fp) != NULL) {
		scan = line;

		/* Scan for SSI tags
		 */
		while ((tag_begin = strstr(scan, "<!--#")) != NULL) {
			if ((tag_end = strstr(tag_begin, "-->")) != NULL) {

				*tag_begin = *tag_end = '\0';
				printf("%s", scan);
				scan = tag_end + 3;

				command = remove_spaces(tag_begin + 5);

				/* Scan for parameters
				 */
				if ((param_key = strchr(command, ' ')) != NULL) {
					*(param_key++) = '\0';

					while (*param_key == ' ') {
						param_key++;
					}

					error_found = false;
					while (*param_key != '\0') {
						if ((param_value = strstr(param_key, "=\"")) == NULL) {
							param = remove_parameters(param);
							error_found = true;
							break;
						}

						*param_value = '\0';
						param_value += 2;

						if ((param_end = strchr(param_value, '"')) == NULL) {
							param = remove_parameters(param);
							error_found = true;
							break;
						}

						*param_end = '\0';

						/* Add parameter to list
						 */
						if ((new = (t_parameter*)malloc(sizeof(t_parameter))) == NULL) {
							print_error(config, "error allocating memory for parameter");
							return -1;
						}
						if (param == NULL) {
							param = new;
						} else {
							last->next = new;
						}
						last = new;

						new->key = param_key;
						new->value = param_value;
						new->next = NULL;

						param_key = param_end + 1;
						while (*param_key == ' ') {
							param_key++;
						}
					}

					if (error_found) {
						continue;
					}
				}

				/* Execute SSI command
				 */
				execute_ssi_command(config, command, param);
				param = remove_parameters(param);
			} else {
				break;
			}
		}

		printf("%s", scan);
	}

	fclose(fp);

	return 0;
}

/* Main
 */
int main(int argc, char *argv[]) {
	t_config *config;

	if (argc <= 1) {
		printf("Missing parameter.\n");
		return EXIT_FAILURE;
	} else if (strcmp(argv[1], "-h") == 0) {
		/* Help
		 */
		printf("Usage: %s <SSI file>\n", argv[0]);
		return EXIT_SUCCESS;
	} else if (strcmp(argv[1], "-v") == 0) {
		/* Version
		 */
		printf("SSI-CGI v"VERSION"\n");
		return EXIT_SUCCESS;
	}

	if ((config = get_configuration()) == NULL) {
		print_error(NULL, "error getting configuration");
		return EXIT_FAILURE;
	}

	printf("Content-Type: text/html\r\n\r\n");

	if (process_ssi_file(config, argv[1]) == -1) {
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}
